btree/btree.c

```diff
@@ -2112,21 +2126,22 @@ static int64_t node_free(struct node_op *op, struct nd *node,
 	int shift = node->n_type->nt_shift(node);
 
 	m0_rwlock_write_lock(&list_lock);
+	node_lock(node);
 	node->n_ref--;
 	node->n_be_node_valid = false;
+	op->no_addr = node->n_addr;
+	m0_free_aligned(segaddr_addr(&op->no_addr), 1ULL << shift, shift);
 
-	if (node->n_ref == 0) {
+	if (node->n_ref == 0 && node->n_txref == 0) {
 		ndlist_tlink_del_fini(node);
+		node_unlock(node);
 		m0_rwlock_fini(&node->n_lock);
-		op->no_addr = node->n_addr;
-		node->n_type->nt_fini(node, tx);
 		m0_free(node);
 		m0_rwlock_write_unlock(&list_lock);
-		m0_free_aligned(segaddr_addr(&op->no_addr), 1ULL << shift,
-				shift);
 		/** Capture in transaction */
```

Comment:
The comment can be removed from here. We should capture in transaction at line 2133.